File: frontend/app/src/main/java/com/capstone/whereigo/MapDataAdapter.java
package com.capstone.whereigo;

import android.app.AlertDialog;
import android.content.Context;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.capstone.whereigo.databinding.ItemMapDataBinding;

import java.io.File;
import java.util.List;

public class MapDataAdapter extends RecyclerView.Adapter<MapDataAdapter.ViewHolder> {

    private final Context context;
    private final List<MapData> mapList;

    public MapDataAdapter(Context context, List<MapData> mapList) {
        this.context = context;
        this.mapList = mapList;
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        public final ItemMapDataBinding binding;

        public ViewHolder(ItemMapDataBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        ItemMapDataBinding binding = ItemMapDataBinding.inflate(inflater, parent, false);
        return new ViewHolder(binding);
    }

    private boolean deleteFileOrFolder(File file) {
        if (file.isDirectory()) {
            File[] children = file.listFiles();
            if (children != null) {
                for (File child : children) {
                    deleteFileOrFolder(child);
                }
            }
        }
        return file.delete();
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        MapData mapData = mapList.get(position);

        holder.binding.tvFileName.setText(mapData.getFileName());
        holder.binding.tvFileSize.setText(mapData.getFileSize());
        holder.binding.tvDate.setText(mapData.getSaveDate());

        holder.binding.btnDelete.setOnClickListener(v -> {
            new AlertDialog.Builder(context)
                    .setTitle("삭제 확인")
                    .setMessage(mapData.getFileName() + "을(를) 삭제하시겠습니까?")
                    .setPositiveButton("삭제", (dialog, which) -> {
                        // FileDownloader와 동일한 구조에 맞춤:
                        File file = new File(context.getExternalFilesDir(null), "Maps/" + mapData.getFileName());
                        boolean deleted = false;

                        Log.d("DeleteCheck", "삭제 대상 경로: " + file.getAbsolutePath());

                        if (file.exists()) {
                            deleted = deleteFileOrFolder(file);
                            Log.d("DeleteCheck", "삭제 결과: " + deleted);
                        } else {
                            Log.d("DeleteCheck", "파일이 존재하지 않음");
                        }

                        // 리스트에서 제거
                        mapList.remove(position);
                        notifyItemRemoved(position);
                        notifyItemRangeChanged(position, mapList.size());

                        if (deleted) {
                            Toast.makeText(context, mapData.getFileName() + "이 삭제되었습니다.", Toast.LENGTH_SHORT).show();
                        } else {
                            Toast.makeText(context, mapData.getFileName() + " 파일 삭제 실패 (목록에서만 제거됨)", Toast.LENGTH_SHORT).show();
                        }
                    })
                    .setNegativeButton("취소", null)
                    .show();
        });
    }

    @Override
    public int getItemCount() {
        return mapList.size();
    }
}

File: frontend/app/src/main/java/com/capstone/whereigo/MapDataAdapter.kt
package com.capstone.whereigo

import android.app.AlertDialog
import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.capstone.whereigo.databinding.ItemMapDataBinding

class MapDataAdapter(
    private val context: Context,
    private val mapList: MutableList<MapData> // 리스트를 MutableList로 변경 (삭제를 위해)
) : RecyclerView.Adapter<MapDataAdapter.ViewHolder>() {

    // ViewHolder에서 ViewBinding을 사용
    class ViewHolder(private val binding: ItemMapDataBinding) : RecyclerView.ViewHolder(binding.root) {
        val fileName = binding.tvFileName
        val fileSize = binding.tvFileSize
        val saveDate = binding.tvDate
        val btnFavorite = binding.btnFavorite
        val btnDelete = binding.btnDelete
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = ItemMapDataBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val mapData = mapList[position]

        holder.fileName.text = mapData.fileName
        holder.fileSize.text = mapData.fileSize
        holder.saveDate.text = mapData.saveDate

        holder.btnFavorite.setImageResource(
            if (mapData.isFavorite) R.drawable.ic_favorite_selected else R.drawable.ic_favorite
        )

        // 삭제 버튼 클릭 이벤트
        holder.btnDelete.setOnClickListener {
            AlertDialog.Builder(context).apply {
                setTitle("삭제 확인")
                setMessage("${mapData.fileName}을(를) 삭제하시겠습니까?")
                setPositiveButton("삭제") { _, _ ->
                    mapList.removeAt(position)
                    notifyItemRemoved(position)
                    notifyItemRangeChanged(position, mapList.size)  // 삭제된 이후의 아이템 position을 갱신
                    Toast.makeText(context, "${mapData.fileName}이 삭제되었습니다.", Toast.LENGTH_SHORT).show()
                }
                setNegativeButton("취소", null)
                show()
            }
        }

        // 즐겨찾기 버튼 클릭 이벤트
        holder.btnFavorite.setOnClickListener {
            mapData.isFavorite = !mapData.isFavorite  // 즐겨찾기 상태 변경

            holder.btnFavorite.setImageResource(
                if (mapData.isFavorite) R.drawable.ic_favorite_selected else R.drawable.ic_favorite
            )

            // 토스트 메시지 출력
            val message = if (mapData.isFavorite) {
                "${mapData.fileName}이 즐겨찾기에 추가되었습니다."
            } else {
                "${mapData.fileName}이 즐겨찾기에서 제거되었습니다."
            }
            Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
        }
    }

    override fun getItemCount() = mapList.size
}
